<a href="https://colab.research.google.com/github/20rashmi128/NYC2019-Airbnb-Bookings-Analysis/blob/main/Notebook_Airbnb_Bookings_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Airbnb Bookings Analysis (EDA)**

Since 2008, guests and hosts have used Airbnb to expand on traveling possibilities and present a more unique, personalized way of experiencing the world. Today, Airbnb became one of a kind service that is used and recognized by the whole world. Data analysis on millions of listings provided through Airbnb is a crucial factor for the company. These millions of listings generate a lot of data - data that can be analyzed and used for security, business decisions, understanding of customers' and providers' (hosts) behavior and performance on the platform, guiding marketing initiatives, implementation of innovative additional services and much more.

This dataset has around 49,000 observations in it with 16 columns and it is a mix between categorical and numeric values.

Explore and analyze the data to discover key understandings (not limited to these) such as :
  What can we learn about different hosts and areas?
  What can we learn from predictions? (ex: locations, prices, reviews, etc)
  Which hosts are the busiest and why?
  Is there any noticeable difference of traffic among different areas and what could be the reason for it?

# 1. Basic Information about Dataset

### 1.(a) Imported relevant python libraries and then loaded the dataset csv file using Pandas.

In [1]:
# Importing relevant libraries.

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Mounted the google drive 
# creating a variable to store path address of required dataset file and then, reading csv file using pandas to a dataframe.

path = '/content/drive/MyDrive/EDA_AirBnB/Copy of Airbnb NYC 2019.csv'
df = pd.read_csv(path)

### 1.(b) Getting basic details/information about the data, its columns and respective data types.


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48879 non-null  object 
 2   host_id                         48895 non-null  int64  
 3   host_name                       48874 non-null  object 
 4   neighbourhood_group             48895 non-null  object 
 5   neighbourhood                   48895 non-null  object 
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  object 
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     

In [4]:
df.columns

Index(['id', 'name', 'host_id', 'host_name', 'neighbourhood_group',
       'neighbourhood', 'latitude', 'longitude', 'room_type', 'price',
       'minimum_nights', 'number_of_reviews', 'last_review',
       'reviews_per_month', 'calculated_host_listings_count',
       'availability_365'],
      dtype='object')

In [5]:
for column in df.columns:
  print("Column",' "',column,'" ', "is of -", type(df.loc[0,column]), "data type")

Column  " id "  is of - <class 'numpy.int64'> data type
Column  " name "  is of - <class 'str'> data type
Column  " host_id "  is of - <class 'numpy.int64'> data type
Column  " host_name "  is of - <class 'str'> data type
Column  " neighbourhood_group "  is of - <class 'str'> data type
Column  " neighbourhood "  is of - <class 'str'> data type
Column  " latitude "  is of - <class 'numpy.float64'> data type
Column  " longitude "  is of - <class 'numpy.float64'> data type
Column  " room_type "  is of - <class 'str'> data type
Column  " price "  is of - <class 'numpy.int64'> data type
Column  " minimum_nights "  is of - <class 'numpy.int64'> data type
Column  " number_of_reviews "  is of - <class 'numpy.int64'> data type
Column  " last_review "  is of - <class 'str'> data type
Column  " reviews_per_month "  is of - <class 'numpy.float64'> data type
Column  " calculated_host_listings_count "  is of - <class 'numpy.int64'> data type
Column  " availability_365 "  is of - <class 'numpy.int64'

A) Observations:-


*   Total 16 columns, out of which 10 are numerical columns (either integer type or float type data) and remaining 6 columns have string type data.
*   Total data/rows available is 48,895 entries.

In [6]:
#  Going through first 5 rows of dataset
df.head()

#  Observations:
#  One row entry has 0 as number of reviews and corresponding last review & reviews per month are missing values.
#  One row entry has 0 availability days for booking.
#  last review column contains the date of the last review. Therefore the same needs to converted to datetime format

id  ... availability_365
0  2539  ...              365
1  2595  ...              355
2  3647  ...              365
3  3831  ...              194
4  5022  ...                0

[5 rows x 16 columns]

In [7]:
#  checking the last 5 rows of dataset
df.tail()

# Observations -
# All last 5 entries have 0 number of reviews and have missing values for columns "last_review" and "reviews_per_month".
# check the logical deduction that 0 reviews must be accompanied by NaN last review and NaN reviews_per_months. 
# If yes, then these missing values are natural and so, need no treatment.

id  ... availability_365
48890  36484665  ...                9
48891  36485057  ...               36
48892  36485431  ...               27
48893  36485609  ...                2
48894  36487245  ...               23

[5 rows x 16 columns]

### 1.(c) Renaming few columns for better clarity

In [8]:
df.rename(columns = {'id':'property_id','name':'property_name','price':'price_dollar'}, inplace = True)
df.head(2)

property_id  ... availability_365
0         2539  ...              365
1         2595  ...              355

[2 rows x 16 columns]

### 1.(d) Descriptive statistical analysis of Numerical type columns/features.

In [9]:
# checking the descriptive stats such as mean, mode etc. for numerical columns.
df.describe()

# Observations:-
# 1. reviews_per_month seems to have few missing values.
# 2. Rest all 9 numerical columns have no missing value.
# 3. Minimum value in price_dollar column is 0.
# 4. There is/are properties whose number of reviews is 0 (i.e., no reviews present).
# 5. There is/are properties whose days' availability for booking is 0 day. (Explain)

property_id  ...  availability_365
count  4.889500e+04  ...      48895.000000
mean   1.901714e+07  ...        112.781327
std    1.098311e+07  ...        131.622289
min    2.539000e+03  ...          0.000000
25%    9.471945e+06  ...          0.000000
50%    1.967728e+07  ...         45.000000
75%    2.915218e+07  ...        227.000000
max    3.648724e+07  ...        365.000000

[8 rows x 10 columns]

In [10]:
print("Total number of unique listings:", len(df))

Total number of unique listings: 48895


Observation:-
1) Each row/data entry has been uniquely defined by unique id.

# 2. Data Preparation & Cleaning

## Steps followed:-
   2. Checking for Missing values and Unique Values of each column.
   3. Handling Missing Values
   4. Checking duplicate values
   5. Checking Outlierrs and handling outliers (if any).
   6. If any column/feature seems to be of zero importance, drop those columns.

# 1. Converting the last_review_dates from string data type to datetime without timestamp.

In [11]:
#  Conerting the last_review column data to datetime format without timestamp.

# Creating new column with last_review data converted to datetime with timestamp format.
df['date_col'] = pd.to_datetime(df.last_review)

# Creating new column with just the date part without timestamp.
df['last_review_date'] = df['date_col'].dt.date
df['last_review_date']

0        2018-10-19
1        2019-05-21
2               NaT
3        2019-07-05
4        2018-11-19
            ...    
48890           NaT
48891           NaT
48892           NaT
48893           NaT
48894           NaT
Name: last_review_date, Length: 48895, dtype: object

In [12]:
df.drop(['date_col'], axis =1, inplace = True)

# 2. Checking and Handling Missing Values

In [13]:
# checking for missing values against all columns of dataset
df.isnull().sum()

property_id                           0
property_name                        16
host_id                               0
host_name                            21
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price_dollar                          0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
last_review_date                  10052
dtype: int64

In [14]:
df[df.property_name.isnull()]

property_id property_name  ...  availability_365 last_review_date
2854       1615764           NaN  ...               362              NaT
3703       2232600           NaN  ...               341       2015-06-08
5775       4209595           NaN  ...                 0       2015-01-01
5975       4370230           NaN  ...                 0       2016-01-02
6269       4581788           NaN  ...                 0              NaT
6567       4756856           NaN  ...                 0              NaT
6605       4774658           NaN  ...                 0              NaT
8841       6782407           NaN  ...                 0              NaT
11963      9325951           NaN  ...                 0       2016-01-05
12824      9787590           NaN  ...                 0              NaT
13059      9885866           NaN  ...                 0              NaT
13401     10052289           NaN  ...               362       2016-08-18
15819     12797684           NaN  ...                 0              NaT
16071     12988898           NaN  ...               365              NaT
18047     14135050           NaN  ...                 0              NaT
28889     22275821           NaN  ...                 0       2018-08-13

[16 rows x 17 columns]

# Properties with 0 days of availabilty for booking purpose.

In [15]:
#  checking the number of entries whose availability of days for booking out of 365 days is zero. 
# So, these properties are listed/active on Airbnb, but are not available for booking the year the data was collected.
print( "The number of AirBnB properties which are not available for booking this year is", len(df[df['availability_365']==0]))

The number of AirBnB properties which are not available for booking this year is 17533


# Properties having zero Reviews

In [16]:
#  Getting the name and total count of those properties, for which users have not given reviews.
null_rvw_df = df[df['number_of_reviews']==0]
print("The below",len(null_rvw_df), "properties have zero reviews:", '\n', null_rvw_df['property_name'].unique())

The below 10052 properties have zero reviews: 
 ['THE VILLAGE OF HARLEM....NEW YORK !' 'Huge 2 BR Upper East  Cental Park'
 'Magnifique Suite au N de Manhattan - vue Cloitres' ...
 'Sunny Studio at Historical Neighborhood'
 '43rd St. Time Square-cozy single bed'
 "Trendy duplex in the very heart of Hell's Kitchen"]


# 3. Data Exploration and Visulaization

## Columns to analyse:-
1. Duplicate property names.
2. Host_id, host_name and property_name
3. Neighbourhood area and location 

# 1. Duplicate Property Names

In [17]:
print("#unique values in 'property_id' column:", df.property_id.nunique())
print("#unique values in 'property_name' column:", df.property_name.nunique())

#unique values in 'property_id' column: 48895
#unique values in 'property_name' column: 47905


In [18]:
duplicate_property = df[df.duplicated(subset=['property_name'])]
print("The number of properting listings' names which are duplicate is",duplicate_property.property_name.nunique())
duplicate_property[3:12]

The number of properting listings' names which are duplicate is 645


property_id  ... last_review_date
661       250537  ...              NaT
669       253471  ...       2018-11-06
670       253475  ...       2019-06-24
674       253800  ...       2019-04-25
675       253803  ...       2019-06-22
676       253806  ...       2019-07-02
677       253811  ...       2019-07-03
678       253815  ...       2019-06-29
679       253828  ...       2018-07-26

[9 rows x 17 columns]

# 2. Information regarding Duplicate Hosts and thier multiple Listings

In [19]:
#  missing values in host_id, host_name
print("Number of missing values in host_id column:", df.host_id.isnull().sum())
print("Number of unique values in 'host_id' column:", df.host_id.nunique())
print(" ")
print("Number of missing values in host_name column:", df.host_name.isnull().sum())
print("Number of unique values in 'host_name' column:", df.host_name.nunique())

Number of missing values in host_id column: 0
Number of unique values in 'host_id' column: 37457
 
Number of missing values in host_name column: 21
Number of unique values in 'host_name' column: 11452


In [20]:
duplicate_hosts = df[df['host_id'].duplicated()]

print("The number of Duplicate hosts are :", duplicate_hosts['host_id'].nunique())
print("The number of Duplicate host names are :", duplicate_hosts['host_name'].nunique())
print("The total proerties managed by these duplicate hosts are:", len(duplicate_hosts))
print(" ")
print("05 Duplicate hosts are :")
duplicate_hosts[:5]

The number of Duplicate hosts are : 5154
The number of Duplicate host names are : 2942
The total proerties managed by these duplicate hosts are: 11438
 
05 Duplicate hosts are :


property_id  ... last_review_date
22         8025  ...       2019-01-01
23         8110  ...       2019-07-02
33         9783  ...       2019-07-01
35        10962  ...       2019-06-21
39        12048  ...       2019-07-05

[5 rows x 17 columns]

In [21]:
df[df.host_id == 7549]

property_id  ... last_review_date
9            5238  ...       2019-06-09
39          12048  ...       2019-07-05
4767      3373030  ...       2019-06-25
5778      4215595  ...       2019-05-19

[4 rows x 17 columns]

In [22]:
df[df.host_id == 32294]

property_id  ... last_review_date
30          9668  ...       2019-06-01
33          9783  ...       2019-07-01
100        22918  ...       2019-01-03

[3 rows x 17 columns]

In [23]:
df[df.host_id == 9744]

property_id  ... last_review_date
12          5803  ...       2019-06-24
35         10962  ...       2019-06-21
192        50447  ...       2019-06-22

[3 rows x 17 columns]

In [24]:
df[df.host_id == 22486]

property_id  ... last_review_date
21           8024  ...       2019-07-01
22           8025  ...       2019-01-01
23           8110  ...       2019-07-02
476        167222  ...       2018-08-24
588        222054  ...       2018-09-15
1783       801626  ...       2019-05-26

[6 rows x 17 columns]

In [25]:
print(len(df[df.host_id == 417504]))
print(df[df.host_id == 417504].calculated_host_listings_count.unique())

28
[28]


Observations:
 1. 5,154 number of hosts are managing multiple properties (i.e., in total 11,438 properties).
 2. Most hosts out of above 5,154 hosts, such as Lisel, Ben, Laurie etc. have enlisted different category of rooms/flats available in their property as separate units/property for booking separately. This also matches the column "calculated_host_listings_count", which is basically the total number of properties listed by a given host.

# 3. Neighborhood Area and locations

In [26]:
print("#unique values in 'neighbourhood_group' column:", df.neighbourhood_group.nunique())
print("#unique values in 'neighbourhood' column:", df.neighbourhood.nunique())

#unique values in 'neighbourhood_group' column: 5
#unique values in 'neighbourhood' column: 221


# 3. Questions and Answers

# 4. Summary:

  Summary should be crisp and pertinent. The outcome of the project and the method used to reach the outcome should be specified. It should be articulate, well-versed, and apposite. Also, add the link to the GitHub project repository.

Insights:-
  1. 
##  2. 

## Any future project ideas:-

# **Evaluation criteria (For reference only):-**

FINAL CHECK

1. Exploration, namely: head, tail, summary, data dictionary - 5
2. Looking for and handling NaN/ Null/ Missing Values - 5
3. Trying to get some conclusion from data, correlation, trends and making sense in doing so - 10
4. Accomplish various milestones given in the problem statement - 10
5. Using visualization (atleast five different types) for presenting the EDA - 15
6. Final summary of conclusion - 5
7. Commented code - 5
8. Proper output formatting - 2.5
9. Modularity of code - 2.5
10. Presentation Slideshow - 5
11. Presentation video - 15
12. Fluency and Grammatical Accuracy in PPT and Video - 5
13. GitHub Commits - 5
14. Summary and Technical Documentation - 10

